petri/src/test.rs

```diff
@@ -104,14 +106,52 @@ impl Test {
             resolve(&name, self.requirements()).context("failed to resolve artifacts")?;
         let output_dir = artifacts.get(petri_artifacts_common::artifacts::TEST_LOG_DIRECTORY);
         let logger = try_init_tracing(output_dir).context("failed to initialize tracing")?;
-        self.test.0.run(
-            PetriTestParams {
-                test_name: &name,
-                logger: &logger,
-                output_dir,
-            },
-            &artifacts,
-        )
+
+        // Catch test panics in order to cleanly log the panic result. Without
+        // this, `libtest_mimic` will report the panic to stdout and fail the
+        // test, but the details won't end up in our per-test JSON log.
+        let r = catch_unwind(AssertUnwindSafe(|| {
+            self.test.0.run(
+                PetriTestParams {
+                    test_name: &name,
+                    logger: &logger,
+                    output_dir,
+                },
+                &artifacts,
+            )
+        }));
+        let r = r.unwrap_or_else(|err| {
+            // The error from `catch_unwind` is almost always either a
+            // `&str` or a `String`, since that's what `panic!` produces.
+            let msg = err
+                .downcast_ref::<&str>()
+                .copied()
+                .or_else(|| err.downcast_ref::<String>().map(|x| x.as_str()));
+
+            let err = if let Some(msg) = msg {
+                anyhow::anyhow!("test panicked: {msg}")
+            } else {
+                anyhow::anyhow!("test panicked")
```

Comment:
Is it worth trying to downcast to dyn Debug if it's not either string type?

Reply:
We should at least put some text here like "non-string result found, this is unexpected" or something

Reply:
You can only downcast to a sized type. This is fundamental--there's no vtable anywhere from which to manifest a `dyn Debug`.

Reply:
Not even the original type name is available, only the type ID, so there's really nothing user-friendly to trace. Note that Rust doesn't have any clever mechanism internally for this either--if you use `panic_any(3)`, Rust will just print out `Box<dyn Any>` as the panic message.

Reply:
Fun. Well we can definitely print out a message saying why there's no message at least, even if we never expect to see it.

Reply:
```suggestion
                anyhow::anyhow!("test panicked (unknown payload type)")
```